### PR TITLE
BUG: Invalid dtypes comparison should not raise `TypeError`

### LINKED ISSUE #19187: dtype comparison return False
```
dtype = pd.date_range("2016-01-01", periods=2, tz="UTC").dtype

>>> dtype == np.dtype("datetime64[ns]")
False

>>> np.dtype("datetime64[ns]") == dtype
TypeError: Cannot interpret 'datetime64[ns, UTC]' as a data type
```

It'd be nice if this returned False instead of raising.

### PR DESCRIPTION
**Pull request by ganesh-k13**
### Invalid dtypes comparison
Changes behavior of comparison to not raise `TypeError` when one of the operands cannot be converted to valid dtype.

### Example:
```py
>>> dtype = pd.date_range("2016-01-01", periods=2, tz="UTC").dtype
>>> dtype == np.dtype("datetime64[ns]")
False
>>> np.dtype("datetime64[ns]") == dtype
False

# Before:
>>> np.dtype("datetime64[ns]") == dtype
TypeError: Cannot interpret 'datetime64[ns, UTC]' as a data type
```

resolves #19187

### CODE REVIEW

File: numpy/core/src/multiarray/descriptor.c [hunk 1]
```diff
@@ -3228,7 +3228,22 @@ arraydescr_richcompare(PyArray_Descr *self, PyObject *other, int cmp_op)
...
+            case Py_LT:
+            case Py_LE:
+            case Py_EQ:
+            case Py_GT:
+            case Py_GE:
+                Py_RETURN_FALSE;
+            case Py_NE:
+                Py_RETURN_TRUE;
+            default:
+                Py_RETURN_NOTIMPLEMENTED;
```
Comment (seberg) [inline numpy/core/src/multiarray/descriptor.c:3245]: My guess is, we should return `NotImplemented` in all of the cases.

File: numpy/core/tests/test_dtype.py [hunk 2]
```diff
@@ -88,6 +88,19 @@ def test_invalid_types(self):
...
+        [('==', False), ('!=', True),
+        ('>', False), ('<', False),
+        ('>=', False), ('<=', False)])
+    def test_richcompare_invalid_dtype(self, operator, expected):
+        # Make sure objects that cannot be converted to valid
+        # dtypes results in False when compared to valid dtypes.
+        # Here 7 cannot be converted to dtype. No exceptions should be raised
+
+        assert eval(f"np.dtype(np.int32) {operator} 7") == expected,\
+                f"dtype richcompare failed for {operator}"
```
Comment (seberg) [inline numpy/core/tests/test_dtype.py:102]: I suppose you used eval, because you were not aware of `import operator; operator.eq`, etc.?

Comment (ganesh-k13) [reply numpy/core/tests/test_dtype.py:102]: Oh thanks! I did not know of this. Shall replace it.

**Comment by BvB93** [comment]
I guess we can remove these two lines as well in that case:
https://github.com/numpy/numpy/blob/341736a4e5f0023c8352a553b84d59ba9dd3356d/numpy/__init__.pyi#L1084-L1085

File: numpy/core/src/multiarray/descriptor.c (hunk 1 above)
Comment (ganesh-k13) [reply numpy/core/src/multiarray/descriptor.c:3245]: Returning `NotImplemented` gives us a `TypeError` again.

Comment (ganesh-k13) [reply numpy/core/src/multiarray/descriptor.c:3245]: Oh, do we catch it above somewhere?

Comment (seberg) [reply numpy/core/src/multiarray/descriptor.c:3245]: I think Python returns False and True for `==` and `!=` when both return NotImplemented?

Comment (ganesh-k13) [reply numpy/core/src/multiarray/descriptor.c:3245]: Oh sorry my bad. Yeah, it works for `==` and `!=`, and raises `TypeError` for others which makes sense.

**Comment by ganesh-k13** [comment]
@BvB93, will it make sense to move the tests to `numpy/typing/tests/`?

**Comment by BvB93** [comment]
> @BvB93, will it make sense to move the tests to `numpy/typing/tests/`?

I don't think so, no. 
The tests in aforementioned directory are all (directly or indirectly) related to static typing,
while the relevant tests are very much related to the runtime behavior of `np.dtype.__eq__`.

**Comment by ganesh-k13** [comment]
Ohh got it. Thanks!

**Review by seberg — APPROVED** [review]
Could test the opposite direction for the operators, but you are already testing the interesting one anyway.

An unconditional `PyErr_Clear()` is always a bit annoying, but unless someone has a good idea soon, I think this is a clear improvement and ready to go in.

(Or does anyone think it should have a release note? seems like a clear bug-fix to me?)

**Comment by mattip** [comment]
Thanks @ganesh-k13. Let's call this a bug-fix and as such needs no specific release note.

**Pull request merged**